pkg/ee/validation/machine/providers.go

```diff
@@ -459,3 +489,234 @@ func getProjectIDOrTenantID(configVarResolver *providerconfig.ConfigVarResolver,
+	}
+	region, err := configVarResolver.GetConfigVarStringValue(rawConfig.RegionID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the value of  alibaba \"region\" from machine config, error: %w", err)
+	}
+	// TODO: verify format of disk size
+	disk, err := configVarResolver.GetConfigVarStringValue(rawConfig.DiskSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the value of alibaba \"disk\" from machine config, error: %w", err)
+	}
+
+	alibabaClient, err := ecs.NewClientWithAccessKey(region, accessKeyID, accessKeySecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client, error: %w", err)
+	}
+
+	requestInstanceTypes := ecs.CreateDescribeInstanceTypesRequest()
+	instanceTypes := []string{instanceType}
+	requestInstanceTypes.InstanceTypes = &instanceTypes
+
+	instTypes, err := alibabaClient.DescribeInstanceTypes(requestInstanceTypes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list instance types, error: %w", err)
+	}
+
+	for _, instType := range instTypes.InstanceTypes.InstanceType {
+		if instType.InstanceTypeId == instanceType {
+			// parse the Alibaba resource requests
+			// memory is in GB and storage is in GB
+			// TODO: CpuCoreCount or Cores
+			cpuReq, err := resource.ParseQuantity(strconv.Itoa(instType.CpuCoreCount))
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse machine cpu request to quantity, error: %w", err)
+			}
+			memReq, err := resource.ParseQuantity(fmt.Sprintf("%fG", instType.MemorySize))
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse machine memory request to quantity, error: %w", err)
+			}
+			storageReq, err := resource.ParseQuantity(fmt.Sprintf("%sG", disk))
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse machine storage request to quantity, error: %w", err)
+			}
+			return NewResourceDetails(cpuReq, memReq, storageReq), nil
+		}
+	}
+
+	return nil, nil
+}
+
+func getHetznerResourceRequirements(ctx context.Context, userClient ctrlruntimeclient.Client, config *types.Config) (*ResourceDetails, error) {
+	rawConfig, err := hetznertypes.GetConfig(*config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get hetzner raw config, error: %w", err)
+	}
+	configVarResolver := providerconfig.NewConfigVarResolver(ctx, userClient)
+
+	serverType, err := configVarResolver.GetConfigVarStringValue(rawConfig.ServerType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the value of \"serverType\" field, error: %w", err)
+	}
+	token, err := configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Token, envHZToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the value of hetzner \"token\" field, error: %w", err)
+	}
+	hClient := hcloud.NewClient(hcloud.WithToken(token))
+
+	size, _, err := hClient.ServerType.Get(ctx, serverType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list hetzner sizes, error: %w", err)
+	}
```

Comment:
WDYT about moving such provider client specific stuff to the provider code in /pkg/handler/common/provider/ . For all the new providers introduced here?